Introduce defer block instructions

crisbeto opened the pull request:
Adds the logic to generate the various `defer` block instructions. The instructions themselves are currently noop. Split up into a few commits to make it easier to review.

Note: I also had to move some code around in the method that generates the `template` instruction so that we can reuse the logic. This will require a TGP.

crisbeto committed:
refactor(compiler): add method for generating template instructions

Moves the logic for creating a `ɵɵtemplate` instruction into a separate method so that it can be reused for `defer` blocks.

crisbeto committed:
refactor(compiler): introduce defer block instructions

Adds the logic for generating `{#defer}`, `{:placeholder}`, `{:loading}` and `{:error}` block instructions in the compiler.

crisbeto commented on packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_deferred/deferred_with_prefetch_triggers.ts:
```diff
@@ -0,0 +1,19 @@
+
+@Component({
+  template: `
+    {#defer prefetch when isVisible() || isReady; prefetch on idle, timer(1337);
```
I was on the fence whether to do one test per trigger or one with all of them. I went with the latter, because the test and the generated code is pretty straightforward and the compliance tests require a lot of boilerplate and produces a lot of code in the `GOLDEN_PARTIAL.js`.

crisbeto commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -1074,56 +1089,142 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
+  }
+
+  private createDeferTriggerInstructions(
+      templateIndex: number, triggers: t.DeferredBlockTriggers, prefetch: boolean) {
```
Continuing the discussion we had over DM: do we pass in the main template index here or the defer block index? My understanding was that we'd need the latter, because this is used to generate the `advance` instructions before `deferWhen` callls.

AndrewKushnir replied:
Agreed, we need defer instruction index here (so that we can calc advance instruction args).

crisbeto commented on packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_deferred/deferred_secondary_blocks_template.js:
```diff
@@ -0,0 +1,55 @@
+    if (rf & 1) {
+      $r3$.ɵɵelementStart(0, "div");
+      $r3$.ɵɵtext(1);
+      $r3$.ɵɵtemplate(3, MyApp_Defer_3_Template, 1, 0);
```
These template instructions are a bit repetitive, but I have an item on my TODO list to add chaining since it'll be relevant for control flow as well.

crisbeto commented:
[Passing TGP](http://test/OCL:555419833:BASE:555455854:1691671306548:ae38245)

mlc-mlapis commented:
@crisbeto You have recently started your work on the `defer` feature and using a block `{#defer ...}` syntax. Does it mean the block syntax was already chosen as the final solution, even for flow controls instead of the `@` one? @alxhub There is no final announcement about it.

JeanMeche commented:
@mlc-mlapis They will change it if the other syntax is chosen ! [cf Alex's comment on Twitter](https://twitter.com/synalx/status/1687136256629805056)

crisbeto commented:
This code will be the same no matter which syntax we end up with. If we go for the other proposal, we only need to change the lexer and update the unit tests.

AndrewKushnir commented on packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_deferred/deferred_with_triggers_template.js:
```diff
@@ -0,0 +1,17 @@
+function MyApp_Template(rf, ctx) {
+  if (rf & 1) {
+    $r3$.ɵɵtemplate(1, MyApp_Defer_1_Template, 1, 1);
+    $r3$.ɵɵtemplate(2, MyApp_DeferPlaceholder_2_Template, 3, 0);
+    $r3$.ɵɵdefer(0, 1, null, null, 2);
```
As discussed offline, it'd be great to have indices of those instruction go incrementally.
```suggestion
    $r3$.ɵɵtemplate(0, MyApp_Defer_1_Template, 1, 1);
    $r3$.ɵɵtemplate(1, MyApp_DeferPlaceholder_2_Template, 3, 0);
    $r3$.ɵɵdefer(2, 1, null, null, 2);
```

AndrewKushnir reviewed (approved):
@crisbeto thanks for the PR! The changes look great, just left a couple minor comments.

crisbeto committed:
refactor(compiler): introduce defer trigger instructions

Adds the logic for generating the instructions for the various deferred triggers.

crisbeto commented:
Caretaker note: the presubmit passes, but there are a few already-broken targets that are preventing the CI from being green. This change also passed a TGP earlier so I don't expect any issues.

pkozlowski-opensource commented:
This PR was merged into the repository by commit 5212b47bbf9f3da890176512296a49c1b98263c8.
